### PR TITLE
[Pytorch Pytest] Use PyTorch to set HIP_VISIBLE_DEVICES (Vers. 2)

### PR DESCRIPTION
**Pull request by HereThereBeDragons**
Replaces #2195

- now respects  if HIP_VISIBLE_DEVICES was set before
- put gpu detection in a subprocess so that HIP_VISIBLE_DEVICES can be set correctly for pytest
-  normalize `gcnArchName` to be able to match with gpu_arch

co-authored @lamikr

### CODE REVIEW

**Comment by HereThereBeDragons** [comment]
not yet tested, @lamikr offered to do it

**Comment by lamikr** [comment]
All the problems that were in previous PR seems to be now resolved. I tested by launching tests both by using a pytorch versions build separately either for gfx1102 or gfx1103 with and without setting the AMDGPU_FAMILY env variable and all combinations of tests worked now ok.

File: external-builds/pytorch/run_linux_pytorch_tests.py [hunk 1]
```diff
@@ -164,107 +169,242 @@ def cmd_arguments(argv: list[str]) -> argparse.Namespace:
...
+
+    try:
+        result = subprocess.run(
+            [sys.executable, "-c", query_script],
+            capture_output=True,
+            text=True,
+            check=True,
+        )
+        visible_gpus = result.stdout.strip().split("\n")
+        return visible_gpus
```
Comment (ScottTodd) [inline external-builds/pytorch/run_linux_pytorch_tests.py:232]: I find it strange to define this script as a string and then run it under `subprocess`. I guess this is to avoid running `import torch` in the same python process?

How about moving the code to a separate file so  it can be run (and tested) on its own, but then still calling it via subprocess here?

File: external-builds/pytorch/run_linux_pytorch_tests.py [hunk 2]
```diff
@@ -164,107 +169,242 @@ def cmd_arguments(argv: list[str]) -> argparse.Namespace:
...
+    if len(gpus) == 0:
+        print("No AMD GPUs detected", file=sys.stderr)
+        sys.exit(1)
+    # Print one GPU per line for easy parsing
+    for gpu in gpus:
+        print(gpu)
+except Exception as e:
+    print(f"ERROR:{e}", file=sys.stderr)
+    sys.exit(1)
+"""
```
Comment (ScottTodd) [inline external-builds/pytorch/run_linux_pytorch_tests.py:265]: Same thought here, though this script is simpler and I do see value in keeping the code that prints and the code that checks for exceptions close... in that case it may be fine.

File: external-builds/pytorch/run_linux_pytorch_tests.py (hunk 1 above)
Comment (HereThereBeDragons) [reply external-builds/pytorch/run_linux_pytorch_tests.py:232]: agree. i will merge it as "as-is" for now and have added it to the todo list: https://github.com/ROCm/TheRock/issues/2176
